streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/StreamLoader.scala

```diff
@@ -0,0 +1,144 @@
+  //                              |                    |
+  //                              |             +------------+                              +---------+
+  //                              |             |            |              failed inserts  | PubSub  |
+  //                              +-------------| good sink  |             +----------------| (failed |
+  //                                            |            |             |                | inserts)|
+  //                                            +------|-----+             |                +---------+
+  //                                                   |                   |
+  //                                                   |             +-----|----+
+  //                                                   |             | BigQuery |
+  //                                                   +-------------| sink     |
+  //                                                                 +-----|----+
+  //                                                                       |
+  //                                                                       |                 +--------+
+  //                                                                       |     insert      |BigQuery|
+  //                                                                       +-----------------|(events)|
+  //                                                                                         +--------+
+
+  private val MaxConcurrency = Runtime.getRuntime.availableProcessors * 16
+  private val GroupByN       = 10
+  private val TimeWindow     = 5.minutes
+
+  implicit val cs: ContextShift[IO] = IO.contextShift(scala.concurrent.ExecutionContext.Implicits.global)
+
+  def run(e: Environment)(implicit cs: ContextShift[IO], c: Concurrent[IO], T: Timer[IO]): IO[ExitCode] = {
+    val source = new PubsubSource(e)(cs, c)
+
+    val eventStream: Stream[IO, Either[BadRow, StreamLoaderRow]] =
+      source.getStream.evalMap(StreamLoaderRow.parse(e.resolverJson))
+
+    for {
+      client <- Bigquery.getClient
+      queue  <- Queue.bounded[IO, Set[ShreddedType]](MaxConcurrency)
+      sinkBadGood = eventStream.observeEither[BadRow, StreamLoaderRow](
+        badSink(e),
+        goodSink(queue, _.inventory, bigquerySink(e, client))
+      )
+      sinkTypes = queue.dequeue.through(aggregateTypes(GroupByN, TimeWindow)).through(typesSink(e))
+      _ <- Stream(sinkTypes, sinkBadGood).parJoin(MaxConcurrency).compile.drain
+    } yield ExitCode.Success
+  }
+
+  // TODO: Maybe sinks can be generalised?
+
+  /**
+    * Sink bad rows to Pub/Sub.
+    * @param e Parsed common environment.
+    * @return A sink that writes bad rows to Pub/Sub.
+    */
+  def badSink(e: Environment): Pipe[IO, BadRow, Unit] =
+    _.parEvalMapUnordered(MaxConcurrency)(badRow =>
+      PubSub.write(PubSubOutput.WriteBadRow(badRow), e.config.projectId, e.config.badRows)
```

Comment:
What happens if the write to PubSub fails ? Is there a retry mechanism ? Is the input event ACKed ?

Reply:
At the moment, the input event is acked as soon as it's read. But Anton has prepared some changes to wait for this to succeed before ack-ing, which I'll incorporate.